core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java

```diff
@@ -57,21 +66,40 @@ public void visit(KubernetesListBuilder list) {
     }
 
     public static ClusterRole createClusterRole(QuarkusControllerConfiguration<?> cri) {
+        final var rules = new LinkedHashMap<String, PolicyRule>();
+        final var clusterRolePolicyRuleFromPrimaryResource = getClusterRolePolicyRuleFromPrimaryResource(cri);
+        final var primaryRuleKey = getKeyFor(clusterRolePolicyRuleFromPrimaryResource);
+        rules.put(primaryRuleKey, clusterRolePolicyRuleFromPrimaryResource);
 
-        Set<PolicyRule> collectedRules = new LinkedHashSet<>();
-        collectedRules.add(getClusterRolePolicyRuleFromPrimaryResource(cri));
-        collectedRules.addAll(getClusterRolePolicyRulesFromDependentResources(cri));
-        collectedRules.addAll(cri.getAdditionalRBACRules());
+        collectAndMergeIfNeededRulesFrom(getClusterRolePolicyRulesFromDependentResources(cri), rules);
+        collectAndMergeIfNeededRulesFrom(cri.getAdditionalRBACRules(), rules);
 
         return new ClusterRoleBuilder()
                 .withNewMetadata()
                 .withName(getClusterRoleName(cri.getName()))
                 .endMetadata()
-                .addAllToRules(mergePolicyRulesOfSameGroupsAndKinds(collectedRules))
+                .addAllToRules(rules.values())
                 .build();
     }
 
-    @NotNull
+    private static void collectAndMergeIfNeededRulesFrom(Collection<PolicyRule> newRules,
+            Map<String, PolicyRule> existingRules) {
+        newRules.forEach(newPolicyRule -> {
+            final var key = getKeyFor(newPolicyRule);
+            existingRules.merge(key, newPolicyRule, (existing, npr) -> {
+                Set<String> verbs1 = new TreeSet<>(existing.getVerbs());
+                verbs1.addAll(npr.getVerbs());
+                existing.setVerbs(verbs1.stream().toList());
```

Comment:
maybe just `new ArrayList<>(verbs1)` or `List.of(verbs1)` would be better so we don't need to create stream